# Assignment 2: Linear Models and Validation Metrics (30 marks total)
### Due: October 10 at 11:59pm

### Name:

### In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.

## Part 1: Classification (14.5 marks total)

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:

### Step 0: Import Libraries

In [76]:
import numpy as np
import pandas as pd

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [78]:
# TO DO: Import spam dataset from yellowbrick library
# TO DO: Print size and type of X and y

# Import the load_spam function from the yellowbrick.datasets module
from yellowbrick.datasets import load_spam

# Load the spam dataset into X and y
X, y = load_spam()

# Print the size and type of X (features)

print("Size of X: ", X.shape)

print("Type of X: ", type(X))

# Print the size and type of y (target labels)
print("Size of y: ", y.shape)

print("Type of y: ", type(y))

Size of X:  (4600, 57)
Type of X:  <class 'pandas.core.frame.DataFrame'>
Size of y:  (4600,)
Type of y:  <class 'pandas.core.series.Series'>


### Step 2: Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [79]:
# TO DO: Check if there are any missing values and fill them in if necessary
# TO DO: Check if there are any missing values and fill them in if necessary

# Calculate the number of missing values in the DataFrame 'X' and store it in 'anyMissingValue'

anyMissingValue = X.isnull().sum()

# Print the count of missing values

print(anyMissingValue)

word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [82]:

# Import the train_test_split function from scikit-learn
from sklearn.model_selection import train_test_split

# Split the dataset into a small training set (X_small_train, y_small_train) and a small test set (X_small_test, y_small_test)


X_small_train, X_small_test, y_small_train, y_small_test  = train_test_split(X, y, random_state=0, train_size=0.05)

### Step 3: Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets:
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

In [83]:
# Import the necessary modules
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# Create a Logistic Regression model with a maximum of 2000 iterations
logreg = LogisticRegression(max_iter=2000)

# Split the entire dataset (X, y) into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

# Create a subset of the features (using the first 2 columns of X)
X_subset = X.iloc[:, :2]

# Split the subset of features and the target variable into training and testing sets
X_train_selective, X_test_selective, y_train_selective, y_test_selective = train_test_split(X_subset, y, random_state=0)

# Create a smaller training dataset (5% of the entire dataset) for faster experimentation
X_small_train, X_small_test, y_small_train, y_small_test = train_test_split(X, y, random_state=0, train_size=0.05)


### Step 4: Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

In [84]:


# Training and evaluating a logistic regression model on the complete dataset
logreg1 = LogisticRegression(max_iter=2000).fit(X_train, y_train)
print("Training set score for complete set: {:.3f}".format(logreg1.score(X_train, y_train)))
print("Validation set score for complete set: {:.3f}".format(logreg1.score(X_test, y_test)))
print('')


# Training and evaluating a logistic regression model on a selective subset of the dataset (contains first two columns)
logreg2 = LogisticRegression(max_iter=2000).fit(X_train_selective, y_train_selective)
print("Training set score for set(contains first two columns): {:.3f}".format(logreg2.score(X_train_selective, y_train_selective)))
print("Validation set score for set(contains first two columns): {:.3f}".format(logreg2.score(X_test_selective, y_test_selective)))
print('')


# Training and evaluating a logistic regression model on a smaller subset of the dataset (contains 5% of the data)
logreg3 = LogisticRegression(max_iter=2000).fit(X_small_train, y_small_train)
print("Training set score for set(that contains 5 percentage of the data): {:.3f}".format(logreg3.score(X_small_train, y_small_train)))
print("Validation set score for set(that contains 5 percentage of the data): {:.3f}".format(logreg3.score(X_small_test, y_small_test)))

Training set score for complete set: 0.929
Validation set score for complete set: 0.937

Training set score for set(contains first two columns): 0.608
Validation set score for set(contains first two columns): 0.613

Training set score for set(that contains 5 percentage of the data): 0.943
Validation set score for set(that contains 5 percentage of the data): 0.906


### Step 5: Visualize Results (4 marks)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [85]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT

# Importing LogisticRegression from scikit-learn and creating a DataFrame to store and display model evaluation results

from sklearn.linear_model import LogisticRegression

data = {

  'Data size' : [len(X_train), len(X_train_selective), len(X_small_train)],

  'training accuracy' : [logreg1.score(X_train, y_train), logreg2.score(X_train_selective, y_train_selective), logreg3.score(X_small_train, y_small_train)],

  'validation accuracy' : [logreg1.score(X_test, y_test), logreg2.score(X_test_selective, y_test_selective), logreg3.score(X_small_test, y_small_test)]

}

results = pd.DataFrame(data)

print(results)


   Data size  training accuracy  validation accuracy
0       3450           0.928696             0.937391
1       3450           0.608406             0.613043
2        230           0.943478             0.905721


### Questions (4 marks)
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

*YOUR ANSWERS HERE*

"Presenting potential security hazards for the user.

Response 1 -

 Full Dataset (3450 samples): Training Accuracy: 92.87% Validation Accuracy: 93.83%. Initial Two Columns of Data (3450 samples): Training Accuracy: 60.84% Validation Accuracy: 61.30%.Just 5% of the Data (230 samples): Training Accuracy: 94.78% Validation Accuracy: 90.66%

I noticed that the training accuracy tends to remain high even with a smaller dataset (comparing 92.87% with the full dataset versus 94.78% with 5% of the data), suggesting that the model can adapt well to the available data. However, validation accuracy may marginally decrease (from 93.83% to 90.66%) with smaller datasets,


but it can still be relatively high if the subset used for validation is representative. Conversely, utilizing a limited subset of features (the first two columns) can result in a significant decrease in accuracy (Training Accuracy: 60.84% and Validation Accuracy: 61.30%), highlighting the importance of careful feature selection.

Response 2 -


 False Positive: A false positive occurs when the spam filter mistakenly labels a legitimate email as spam. False Negative: A false negative takes place when the spam filter wrongly categorizes a spam email as legitimate.


 In my view, the consequence of a false negative is potentially graver. It signifies that unwanted spam emails go unfiltered and might clutter the user's inbox. Some of these spam emails might be phishing attempts or contain malware, presenting security risks to the user."







### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*




A1:  The primary source of the code was the practice example provided by the instructor in D2L. Additionally, I utilized ChatGPT to gain insights into certain instructions.

A2:
1. I began by loading and organizing the data into a feature matrix and a target vector.
2. Next, I checked for any missing values in the dataset.
3. I selected the appropriate model class.
4. I proceeded to instantiate the model and configure the hyperparameters.
5. After that, I trained the model with the data.
6. I used the trained model to make predictions for new data points.
7. Finally, I assessed and compared the accuracy of the model.

A3: I incorporated several prompts such as:

  a. "What is a tuple object, and can you provide an example of a tuple in a dataset?"

   b. "How can I create a subset of a dataset?" I adjusted the code accordingly.

   c. "How can I create a loop to insert data values into the dataframe?"

I had to make code adjustments as I encountered difficulties in extracting a subset of the data.

A4:
 Yes, I encountered some challenges. For instance, I had difficulties with data splitting and understanding how regression models work. To overcome these challenges, I referred to lecture notes and examples available on D2L, which provided the necessary guidance and clarification for me to be successful.



1.   List item
2.   List item



## Part 2: Regression (10.5 marks total)

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library:
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [86]:
# TO DO: Import spam dataset from yellowbrick library
# TO DO: Print size and type of X and y


# Load the concrete dataset using Yellowbrick

from yellowbrick.datasets import load_concrete

# Load the dataset into variables X and y


X, y = load_concrete()

# Print the size and type of the feature matrix (X)

print("Size of X: ", X.shape)

print("Type of X: ", type(X))

# Print the size and type of the target variable (y)

print("Size of y: ", y.shape)

print("Type of y: ", type(y))


Size of X:  (1030, 8)
Type of X:  <class 'pandas.core.frame.DataFrame'>
Size of y:  (1030,)
Type of y:  <class 'pandas.core.series.Series'>


### Step 2: Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [87]:
# TO DO: Check if there are any missing values and fill them in if necessary

# Check for missing values in the DataFrame X
anyMissingValue = X.isnull().sum()

# Print the count of missing values
print(anyMissingValue)

cement    0
slag      0
ash       0
water     0
splast    0
coarse    0
fine      0
age       0
dtype: int64


### Step 3: Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Implement the machine learning model with `X` and `y`

In [88]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0


# Import necessary libraries and modules
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

# Create a Linear Regression model
lr = LinearRegression()

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

# Fit the Linear Regression model to the training data
lr.fit(X_train, y_train)

LinearRegression()

### Step 4: Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [90]:
# TO DO: ADD YOUR CODE HERE

# Calculate and print the Mean Squared Training score

print("Mean Squared Training score: {:.2f}".format(mean_squared_error(y_train, lr.predict(X_train))))

# Calculate and print the Mean Squared Validation score

print("Mean Squared Validation score: {:.2f}".format(mean_squared_error(y_test, lr.predict(X_test))))

# Leave a blank line for clarity

print('')


# Calculate and print the R2 Training score

print("R2 Training score: {:.2f}".format(r2_score(y_train, lr.predict(X_train))))


# Calculate and print the R2 Validation score

print("R2 Valdiation score: {:.2f}".format(r2_score(y_test, lr.predict(X_test))))





Mean Squared Training score: 111.36
Mean Squared Validation score: 95.90

R2 Training score: 0.61
R2 Valdiation score: 0.62


### Step 5: Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [92]:
# TO DO: ADD YOUR CODE HERE

# Define a dictionary 'data' to store training and validation accuracy metrics

data = {

  'Training Accuracy' : [round(mean_squared_error(y_train, lr.predict(X_train)), 2), round(r2_score(y_train, lr.predict(X_train)), 2)],

  'Validation Accuracy' : [round(mean_squared_error(y_test, lr.predict(X_test)), 2), round(r2_score(y_test, lr.predict(X_test)), 2) ]

}

# Create a Pandas DataFrame 'results' from the 'data' dictionary with custom index

results = pd.DataFrame(data, index = ['MSE', 'R2 score'])


# Print the DataFrame 'results' to display the accuracy metrics

print(results)


          Training Accuracy  Validation Accuracy
MSE                  111.36                95.90
R2 score               0.61                 0.62


### Questions (2 marks)
1. Did using a linear model produce good results for this dataset? Why or why not?



Answer:-


In this instance, the performance of the linear model was subpar. The Mean Squared Error (MSE) during training stood at approximately 111.36, while the validation MSE was around 95.90. In terms of R2 scores, which assess how well the model accounts for the data, the values were only about 0.61 for training and 0.62 for validation.

These figures indicate that the linear model encountered challenges in making precise predictions. The elevated MSE values suggest that the model's predictions deviated significantly from the actual values, and the R2 scores signify that it could only elucidate around 60% of the data's variability.

Hence, employing a linear model did not yield favorable outcomes for this dataset. It might be advisable to explore alternative models or strategies in order to enhance accuracy.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*


Answer 1:

The main source of the code was the practice example given by the instructor in D2L. In addition, I used ChatGPT to learn some of the instructions.

Answer 2:
1. Loading and arrange data into feature matrix and target vector
2. Checking any missing values
3. Choose model class
4. Instantiating model and selecting hyperparameters
5.Fitting model to data
6.Predict values for new data
7. Comparing accuracy.

Answer 3 :

  I used some of the prompts like:-What are mean square error and R2 Score and how with these parameters define the model

Answer 4 :   

 Yes, I faced some difficulties like I was not able to understand R2 score then I took help from google as well..


*DESCRIBE YOUR PROCESS HERE*

## Part 3: Observations/Interpretation (3 marks)

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


*ADD YOUR FINDINGS HERE*





Part 1 - Full Data Set vs. First Two Columns:
When using the entire dataset, we observe a notable increase in both training and validation accuracy, demonstrating strong performance. Conversely, when restricting the analysis to just the first two columns, there is a significant decrease in accuracy, pointing to the necessity for more data or more informative features.

Full Data Set vs. 5% of the Data:
Surprisingly, even with just 5% of the dataset, training accuracy remains quite high. Although validation accuracy is slightly lower, it remains at a satisfactory level, suggesting that this smaller subset might contain crucial patterns worth exploring.

Part 2 - "Upon examining the outcomes, a distinct trend becomes evident. The Mean Squared Error (MSE) values for both the training and validation datasets are relatively elevated, with the training data registering around 111.36 and the validation data around 95.90. Furthermore, the R2 scores, which gauge the model's goodness of fit to the data, hover around 0.61 for training and 0.62 for validation. These results align with our classroom discussions. A high MSE signifies the linear model's difficulty in making accurate predictions, and while the R2 scores are not exceedingly low, they indicate that the model could only account for roughly 60% of the variance within the data."


## Part 4: Reflection (2 marks)
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


*ADD YOUR THOUGHTS HERE*

I appreciated the structure of this assignment. When creating it, I had the opportunity to revisit all the elements we've studied. The most demanding aspect was addressing the sections related to MSE and R2 scores.

## Part 5: Bonus Question (4 marks)

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [93]:

# Importing necessary libraries and modules

from sklearn.linear_model import Ridge

from sklearn.model_selection import train_test_split

from sklearn.metrics import mean_squared_error, r2_score

# Creating a Ridge regression model with alpha=1

ridge = Ridge(alpha=1)

# Splitting the data into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)


# Fitting the Ridge model to the training data

ridge.fit(X_train, y_train)

# Printing the R2 score for the training set of the Ridge model


print("R2 Training score for Ridge model: {:.2f}".format(r2_score(y_train, ridge.predict(X_train))))

# Printing the R2 score for the testing set of the Ridge model

print("R2 Valdiation score for Ridge model: {:.2f}".format(r2_score(y_test, ridge.predict(X_test))))



# Importing the Lasso regression model


from sklearn.linear_model import Lasso

# Creating a Lasso regression model with alpha=0.001


lasso = Lasso(alpha=0.001)

# Fitting the Lasso model to the training data


X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)

lasso.fit(X_train, y_train)

# Printing the R2 score for the training set of the Lasso model

print("R2 Training score for Lasso model: {:.2f}".format(r2_score(y_train, lasso.predict(X_train))))

# Printing the R2 score for the testing set of the Lasso model

print("R2 Valdiation score for Lasso model: {:.2f}".format(r2_score(y_test, lasso.predict(X_test))))

R2 Training score for Ridge model: 0.61
R2 Valdiation score for Ridge model: 0.62
R2 Training score for Lasso model: 0.61
R2 Valdiation score for Lasso model: 0.62


*ANSWER HERE*

Answer :
In practice, a favorable R^2 score typically falls in the range of 0.7 to 0.8 or higher, signifying the model's ability to explain a significant portion of the data's variability. Nevertheless, when using the Ridge model with alpha values ranging from 0.01 to 37, its highest attainable R^2 value is only around 0.54, and the same holds true for the Lasso model at alpha = 0.01. However, when comparing these models to Linear Regression, it becomes evident that neither of them yields a satisfactory performance.